homeassistant/components/dsmr/__init__.py

```diff
@@ -38,3 +52,77 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 async def async_update_options(hass: HomeAssistant, entry: ConfigEntry) -> None:
     """Update options."""
     await hass.config_entries.async_reload(entry.entry_id)
+
+
+@callback
+def async_migrate_entity_entry(
+    config_entry: ConfigEntry, entity_entry: er.RegistryEntry
+) -> dict[str, Any] | None:
+    """Migrate DSMR entity entries.
+
+    - Migrates unique ID for sensors based on entity description name to key.
+    """
+
+    # Replace names with keys in unique ID
+    for old, new in (
+        ("Power Consumption", "current_electricity_usage"),
+        ("Power Production", "current_electricity_delivery"),
+        ("Power Tariff", "electricity_active_tariff"),
+        ("Energy Consumption (tarif 1)", "electricity_used_tariff_1"),
+        ("Energy Consumption (tarif 2)", "electricity_used_tariff_2"),
+        ("Energy Production (tarif 1)", "electricity_delivered_tariff_1"),
+        ("Energy Production (tarif 2)", "electricity_delivered_tariff_2"),
+        ("Power Consumption Phase L1", "instantaneous_active_power_l1_positive"),
+        ("Power Consumption Phase L2", "instantaneous_active_power_l2_positive"),
+        ("Power Consumption Phase L3", "instantaneous_active_power_l3_positive"),
+        ("Power Production Phase L1", "instantaneous_active_power_l1_negative"),
+        ("Power Production Phase L2", "instantaneous_active_power_l2_negative"),
+        ("Power Production Phase L3", "instantaneous_active_power_l3_negative"),
+        ("Short Power Failure Count", "short_power_failure_count"),
+        ("Long Power Failure Count", "long_power_failure_count"),
+        ("Voltage Sags Phase L1", "voltage_sag_l1_count"),
+        ("Voltage Sags Phase L2", "voltage_sag_l2_count"),
+        ("Voltage Sags Phase L3", "voltage_sag_l3_count"),
+        ("Voltage Swells Phase L1", "voltage_swell_l1_count"),
+        ("Voltage Swells Phase L2", "voltage_swell_l2_count"),
+        ("Voltage Swells Phase L3", "voltage_swell_l3_count"),
+        ("Voltage Phase L1", "instantaneous_voltage_l1"),
+        ("Voltage Phase L2", "instantaneous_voltage_l2"),
+        ("Voltage Phase L3", "instantaneous_voltage_l3"),
+        ("Current Phase L1", "instantaneous_current_l1"),
+        ("Current Phase L2", "instantaneous_current_l2"),
+        ("Current Phase L3", "instantaneous_current_l3"),
+        ("Max power per phase", "belgium_max_power_per_phase"),
+        ("Max current per phase", "belgium_max_current_per_phase"),
+        ("Energy Consumption (total)", "electricity_imported_total"),
+        ("Energy Production (total)", "electricity_exported_total"),
+    ):
+        old = old.replace(" ", "_")
```

Comment:
If you always want `old` values to have underscores, let's write it down like that. Otherwise this will be calculated at every startup.

Reply:
fixed in [e29097a](https://github.com/home-assistant/core/pull/74895/commits/e29097adb1b04b252cfd7987a91b7141f9148f19)